libraries/ms-common/src/stm32f0xx/spi.c

```diff
@@ -0,0 +1,185 @@
+  if (!s_received) {
+    PWR_EnterSleepMode(PWR_SLEEPEntry_WFI);
+  }
+  return s_rx_buffer;
+}
+
+StatusCode spi_init(uint8_t spi_x, SPISettings *settings) {
+  if (!prv_is_periph_valid(spi_x) || !prv_are_settings_valid(settings)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // Enable peripheral clock
+  if (spi_x == 1) {
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_SPI1, ENABLE);
+  } else {
+    RCC_APB1PeriphClockCmd(RCC_APB1Periph_SPI2, ENABLE);
+  }
+
+  // Note: gpio_init_pin initializes pin clocks as well
+
+  // Initialize MOSI pin
+  GPIOSettings pin_settings = { 0 };
+  pin_settings.direction = GPIO_DIR_OUT;
+  pin_settings.state = GPIO_STATE_LOW;
+  pin_settings.resistor = GPIO_RES_NONE;
+  pin_settings.alt_function = GPIO_ALTFN_0;
+  gpio_init_pin(&settings->mosi_pin, &pin_settings);
+
+  // Initialize SCK pin
+  gpio_init_pin(&settings->sck_pin, &pin_settings);
+
+  // Initialize MISO pin
+  pin_settings.direction = GPIO_DIR_IN;
+  pin_settings.resistor = GPIO_RES_PULLUP;
+  gpio_init_pin(&settings->miso_pin, &pin_settings);
+
+  // Initialize CS pin
+  pin_settings.direction = GPIO_DIR_OUT;
+  pin_settings.state = GPIO_STATE_HIGH;
+  gpio_init_pin(&settings->cs_pin, &pin_settings);
+  s_cs_pin_map[spi_x] = settings->cs_pin;
+
+  // Fill SPI init struct
+  SPI_InitTypeDef spi_init_struct;
+
+  // Default settings
+  spi_init_struct.SPI_Direction = SPI_Direction_2Lines_FullDuplex;
+  spi_init_struct.SPI_Mode = SPI_Mode_Master;
+  spi_init_struct.SPI_NSS = SPI_NSS_Soft;
+  spi_init_struct.SPI_CRCPolynomial = 7;
+  spi_init_struct.SPI_DataSize = SPI_DataSize_8b;
+
+  // Parse clock polarity
+  if (settings->polarity == SPI_CPOL_LOW) {
+    spi_init_struct.SPI_CPOL = SPI_CPOL_Low;
+  } else {
+    spi_init_struct.SPI_CPOL = SPI_CPOL_High;
+  }
+
+  // Parse clock phase
+  if (settings->phase == SPI_CPHA_1EDGE) {
+    spi_init_struct.SPI_CPHA = SPI_CPHA_1Edge;
+  } else {
+    spi_init_struct.SPI_CPHA = SPI_CPHA_2Edge;
+  }
+
+  // Parse Baud Rate prescaler value
+  spi_init_struct.SPI_BaudRatePrescaler = ((settings->baud_rate / 2) << 4)
+                                      + ((settings->baud_rate % 2) << 3);
+
+  // Parse first bit
+  if (settings->first_bit == SPI_FIRSTBIT_MSB) {
+    spi_init_struct.SPI_FirstBit = SPI_FirstBit_MSB;
+  } else {
+    spi_init_struct.SPI_FirstBit = SPI_FirstBit_LSB;
+  }
+
+  Spi_Init(spi_periph_map[spi_x], &spi_init_struct);
+
+  // Enable NVIC for interrupts
+  NVIC_InitTypeDef nvic_init_struct = { 0 };
+  nvic_init_struct.NVIC_IRQChannel = SPI1_IRQn + spi_x;
+  nvic_init_struct.NVIC_IRQChannelPriority = 1; // no idea if this is correct priority
+  nvic_init_struct.NVIC_IRQChannelCmd = ENABLE;
+
+  // Set FIFO threshold for RXNE event
+  SPI_RxFIFOThresholdConfig(spi_periph_map[spi_x], SPI_RxFIFOThreshold_QF);
+
+  // Enable peripheral
+  SPI_Cmd(spi_periph_map[spi_x], ENABLE);
+  return STATUS_CODE_OK;
+}
+
+StatusCode spi_exchange(uint8_t spi_x, uint8_t *tbuf, uint8_t *rbuf,
```

Comment:
Generally use `tx_data/rx_data` and `tx_len/rx_len` or similar. They're more descriptive than `tbuf` and consistent with each other. Kinda weird to use underscores for one variable but not the other. :P 
I usually try to organize function parameters so it goes `(object, input, output)` where `object` is a pointer to the struct, `input` includes any input parameters, and `output` is whenever you're passing in a pointer to storage with the expectation that it will be written to (e.g. receiving data).